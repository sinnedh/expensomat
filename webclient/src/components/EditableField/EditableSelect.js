import React from 'react'
import PropTypes from 'prop-types'

class EditableSelect extends React.Component {
  state = {
    editValue: this.props.value,
    editMode: false
  }

  handleClickEdit(event) {
    event.preventDefault()
    this.setState({editValue: this.props.value, editMode: true})
  }

  handleClickCancel(event) {
    event.preventDefault()
    this.setState({editValue: '', editMode: false})
  }

  handleClickSave(event) {
    event.preventDefault()
    this.setState({editMode: false})
    this.props.onClickSave(this.state.editValue)
  }

  handleChangeValue(event) {
    this.setState({editValue: event.target.value})
  }

  render() {
    if(this.state.editMode) {
      return (
        <span>
          <select className="value" value={this.state.editValue} onChange={e => this.handleChangeValue(e)}>
          {Object.keys(this.props.options).map((key) =>
            <option key={key} value={key}>{this.props.options[key]}</option>
          )}
          </select>
          <input className="save-button" type="button" value="Save" onClick={e => this.handleClickSave(e)} />
          <input className="cancel-button" type="button" value="Cancel" onClick={e => this.handleClickCancel(e)} />
        </span>
      )
    } else {
      return (
        <span>
          {this.props.options[this.props.value]}
          <input className="edit-button" type="button" value="Edit" onClick={e => this.handleClickEdit(e)} />
        </span>
      )
    }
  }
}

EditableSelect.propTypes = {
  options: PropTypes.object,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  onClickSave: PropTypes.func.isRequired,
}

EditableSelect.defaultProps = {
  options: [],
  value: '',
  onClickSave: (value) => {},
}

export default EditableSelect
