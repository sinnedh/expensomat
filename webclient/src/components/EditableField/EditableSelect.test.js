import React from 'react'
import ReactDOM from 'react-dom'
import EditableSelect from './EditableSelect'

const options = {hello: "Hallo", ciao: "Ciao"}

it('matches snapshot when in show mode', () => {
  const wrapper = shallow(<EditableSelect value={"hello"} options={options} />)
  wrapper.setState({editMode: false})
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when in edit mode', () => {
  const wrapper = shallow(<EditableSelect value={"ciao"}  options={options} />)
  wrapper.setState({editMode: true})
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when not isEditable', () => {
  const wrapper = shallow(<EditableSelect value={"hello"} options={options} isEditable={false} />)
  expect(wrapper).toMatchSnapshot()
})

it('calls onUpdateInput when input is updated', () => {
  const wrapper = shallow(<EditableSelect value={"hello"} options={options} />)
  wrapper.setState({editMode: true})
  wrapper.find('select.value').simulate('change', { target: { value: 'ciao' } })
  expect(wrapper.state('editValue')).toBe('ciao')
})

it('shows edit mode when edit button is clicked', () => {
  const wrapper = shallow(<EditableSelect value={"Hello"} options={options} />)
  wrapper.find('input.edit-button').simulate('click', { preventDefault() {} })
  expect(wrapper.find('select.value')).toHaveLength(1)
  expect(wrapper.state('editMode')).toBe(true)
})

it('leaves edit mode when cancel button is clicked', () => {
  const wrapper = shallow(<EditableSelect value={"Hello"} options={options} />)
  wrapper.setState({editMode: true})
  wrapper.find('input.cancel-button').simulate('click', { preventDefault() {} })
  expect(wrapper.find('select.value')).toHaveLength(0)
  expect(wrapper.state('editMode')).toBe(false)
})

it('save button calls onClickSave with input value', () => {
  const onClickSave = jest.fn()
  const wrapper = shallow(
    <EditableSelect
      onClickSave={onClickSave}
      value={"hello"}
      options={options}
    />
  )

  wrapper.setState({editMode: true})
  wrapper.find('input.save-button').simulate('click', { preventDefault() {} })
  expect(onClickSave).toBeCalledWith('hello')
  expect(wrapper.find('select.value')).toHaveLength(0)
})
