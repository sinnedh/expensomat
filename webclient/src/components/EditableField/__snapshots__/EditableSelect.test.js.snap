// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when in edit mode 1`] = `
<span>
  <select
    className="value"
    onChange={[Function]}
    value="ciao"
  >
    <option
      key="hello"
      value="hello"
    >
      Hallo
    </option>
    <option
      key="ciao"
      value="ciao"
    >
      Ciao
    </option>
  </select>
  <input
    className="save-button"
    onClick={[Function]}
    type="button"
    value="Save"
  />
  <input
    className="cancel-button"
    onClick={[Function]}
    type="button"
    value="Cancel"
  />
</span>
`;

exports[`renders when in show mode 1`] = `
<span>
  Hallo
  <input
    className="edit-button"
    onClick={[Function]}
    type="button"
    value="Edit"
  />
</span>
`;
